examples/ssh/opkssh/cli.go

```diff
@@ -0,0 +1,319 @@
+	allowedEmail, allowedPrincipals, err := p.ReadPolicyFile()
+	if err != nil {
+		return err
+	}
+	var claims struct {
+		Email string `json:"email"`
+	}
+	if err := json.Unmarshal(pkt.Payload, &claims); err != nil {
+		return err
+	}
+	if string(claims.Email) == allowedEmail {
+		if slices.Contains(allowedPrincipals, principalDesired) {
+			// Access granted
+			return nil
+		} else {
+			return fmt.Errorf("no policy to allow %s to assume %s, check policy config in %s", claims.Email, principalDesired, p.PolicyFilePath)
+		}
+	} else {
+		return fmt.Errorf("no policy for email %s, allowed email is %s, check policy config in %s", claims.Email, allowedEmail, p.PolicyFilePath)
+	}
+}
+
+// This function is called by the SSH server as the AuthorizedKeysCommand:
+//
+// The following lines are added to /etc/ssh/sshd_config:
+//
+//	AuthorizedKeysCommand /etc/opk/opkssh ver %u %t %k
+//	AuthorizedPrincipalsCommandUser root
+//
+// The parameters specified in the config map the parameters sent to the function below.
+// We prepend "Arg" to specify which ones are arguments sent by sshd. They are:
+//
+//	%u The username (requested principal) - userArg
+//	%t The public key type - typArg - in this case a certificate being used as a public key
+//	%k The base64-encoded public key for authentication - certB64Arg - the public key is also a certificate
+func AuthorizedKeysCommand(userArg string, typArg string, certB64Arg string, policyEnforcer PolicyCheck, op client.OpenIdProvider) (string, error) {
+	cert, err := sshcert.NewFromAuthorizedKey(typArg, certB64Arg)
+	if err != nil {
+		return "", err
+	}
+	if pkt, err := cert.VerifySshPktCert(op); err != nil {
+		return "", err
+	} else if err := policyEnforcer(userArg, pkt); err != nil {
+		return "", err
+	} else {
+		// sshd expects the public key in the cert, not the cert itself.
+		// This public key is key of the CA the signs the cert, in our
+		// setting there is no CA.
+		pubkeyBytes := ssh.MarshalAuthorizedKey(cert.SshCert.SignatureKey)
+		return "cert-authority " + string(pubkeyBytes), nil
+	}
+}
+
+func CheckCert(userDesired string, cert *sshcert.SshCertSmuggler, policyEnforcer PolicyCheck, op client.OpenIdProvider) error {
+	pkt, err := cert.VerifySshPktCert(op)
+	if err != nil {
+		return err
+	}
+	return policyEnforcer(userDesired, pkt)
+}
+
+type PolicyCheck func(userDesired string, pkt *pktoken.PKToken) error
+
+func CreateSSHCert(cxt context.Context, client *client.OpkClient, signer crypto.Signer, alg jwa.KeyAlgorithm, gqFlag bool, principals []string) ([]byte, []byte, error) {
+	pkt, err := client.OidcAuth(cxt, signer, alg, map[string]any{}, gqFlag)
+	cert, err := sshcert.New(pkt, principals)
+	if err != nil {
+		return nil, nil, err
+	}
+	sshSigner, err := ssh.NewSignerFromSigner(signer)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	signerMas, err := ssh.NewSignerWithAlgorithms(sshSigner.(ssh.AlgorithmSigner), []string{ssh.KeyAlgoECDSA256})
```

Comment:
I still have no idea what "Mas" is

Reply:
alright, I think this is supposed to be "MultiAlgorithmSigner" which is incorrect here because you're using an `ssh.AlgorithmSigner`. Maybe `algSigner` or something that's a bit more readable?